src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -13,37 +13,210 @@
+    SHUFFLE_REMOVE_THROW = 16 // shuffle thread that is being removed and needs to throw an exception
+};
+
+/**
+ * Convert a state to a string representation for logging.
+ */
+const char * as_str(thread_state state) {
+  switch(state) {
+    case TASK_RUNNING: return "TASK_RUNNING";
+    case TASK_WAIT_ON_SHUFFLE: return "TASK_WAIT_ON_SHUFFLE";
+    case TASK_BUFN_WAIT_ON_SHUFFLE: return "TASK_BUFN_WAIT_ON_SHUFFLE";
+    case TASK_ALLOC: return "TASK_ALLOC";
+    case TASK_ALLOC_FREE: return "TASK_ALLOC_FREE";
+    case TASK_BLOCKED: return "TASK_BLOCKED";
+    case TASK_BUFN_THROW: return "TASK_BUFN_THROW";
+    case TASK_BUFN_WAIT: return "TASK_BUFN_WAIT";
+    case TASK_BUFN: return "TASK_BUFN";
+    case TASK_SPLIT_THROW: return "TASK_SPLIT_THROW";
+    case TASK_REMOVE_THROW: return "TASK_REMOVE_THROW";
+    case SHUFFLE_RUNNING: return "SHUFFLE_RUNNING";
+    case SHUFFLE_ALLOC: return "SHUFFLE_ALLOC";
+    case SHUFFLE_ALLOC_FREE: return "SHUFFLE_ALLOC_FREE";
+    case SHUFFLE_BLOCKED: return "SHUFFLE_BLOCKED";
+    case SHUFFLE_THROW: return "SHUFFLE_THROW";
+    case SHUFFLE_REMOVE_THROW: return "SHUFFLE_REMOVE_THROW";
+    default: return "UNKNOWN";
+  }
+}
+
+static std::shared_ptr<spdlog::logger> make_logger(std::ostream& stream) {
+  return std::make_shared<spdlog::logger>(
+    "SPARK_RMM", std::make_shared<spdlog::sinks::ostream_sink_mt>(stream));
+}
+
+static std::shared_ptr<spdlog::logger> make_logger() {
+  return std::make_shared<spdlog::logger>(
+    "SPARK_RMM", std::make_shared<spdlog::sinks::null_sink_mt>());
+}
+
+static auto make_logger(std::string const& filename) {
+  return std::make_shared<spdlog::logger>(
+    "SPARK_RMM", std::make_shared<spdlog::sinks::basic_file_sink_mt>(filename, true /*truncate file*/));
+}
+
+/**
+ * The priority of a thread is primarily based off of the task id. The thread id (PID on linux) is
+ * only used as a tie breaker if a task has more than a single thread associated with it.
+ * In Spark a tasks increase sequentially as they are assigned in an application. We want to give
+ * priority to tasks that came first. This is to avoid situations where the first task stays as
+ * the lowest priority task and is constantly retried while newer tasks move to the front of the
+ * line. So a higher task_id should be a lower priority. 
+ *
+ * We also want all shuffle threads to have the highest priority possible. So we assign them
+ * a task id of -1. The problem is overflow on a long, so for the priority of a task the formula
+ * will be MAX_LONG - (task_id + 1).
+ */
+class thread_priority {
+public:
+  thread_priority(long tsk_id, long t_id): task_id(tsk_id), thread_id(t_id) {}
+
+  long get_thread_id() const {
+    return thread_id;
+  }
+
+  long get_task_id() const {
+    return task_id;
+  }
+
+  bool operator<(const thread_priority &other) const {
+    long task_priority = std::numeric_limits<long>::max() - task_id + 1;
```

Comment:
I think you need parens here:  max() - (task_id + 1)
Same in following lines.

Reply:
nit, may be nice to put this formula in a function so that we call that to calculate the priority.